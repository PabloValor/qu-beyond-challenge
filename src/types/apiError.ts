export interface ApiError { 
    message: string
    reason?: string
    level?: number
    trace?: string
}