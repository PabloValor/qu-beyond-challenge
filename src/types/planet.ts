import Film from "./film"
import People  from "./people"

export interface Planet {
    id?: string
    name: string
    diameter: string
    rotation_period: string
    orbital_period: string
    gravity: string
    population: string
    climate: string
    terrain: string
    surface_water: string
    residents: People[]
    films: Film[]
    url: string
    created: string
    edited: string
}

export interface PlanetResponse { 
    count: number
    next: string
    previous: string
    results: Planet[]
}