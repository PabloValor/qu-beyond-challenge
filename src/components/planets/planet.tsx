import { Planet as TPlanet} from "../../types/planet"
import { Badget } from "../util/badget"
import { FaTemperatureHigh } from 'react-icons/fa'
import { IoMdPeople } from 'react-icons/io'
import { Button } from "../util/button"
import slugify from 'slugify'
import { Link } from "react-router-dom"
import { getRandomImageId } from "../../util"

export const displayPopulationNumber = (population: string) => {

    const number = Number(population)

    if(isNaN(number)) return population

    if(number >= 1000000) {
        return `${number / 1000000}M`
    } else if(number >= 1000) { 
        return `${number / 1000}K`
    }
    return number
}


export default function Planet({ planet } : { planet : TPlanet }) { 


    return (
        <div className={`
            p-3
            rounded-md shadow-lg
            cursor-pointer
            w-full md:max-w-[300px]
            bg-[url('/img/bg-galaxy.png')] bg-center bg-cover bg-no-repeat
            relative
            hover:bottom-1
        `}>
            <h2 className="font-bold text-center text-slate-50">{planet.name.toUpperCase()}</h2>

            <Image planet={planet}/>

            <Badget styles="mr-1">
                <FaTemperatureHigh className="inline-block mr-1"/>
                <span data-testid="climate">{planet.climate}</span>
            </Badget>
            <Badget styles="mr-1">
                <IoMdPeople className="inline-block mr-1"/>
                <span data-testid="population" title={`population: ${planet.population}`}>{displayPopulationNumber(planet.population)}</span>
            </Badget>
            
        </div>
    )

    function Image({ planet } : {planet : TPlanet}) { 
        return (
            <Link to={`/planet/${slugify(planet.name, { lower: true })}/${planet.id}`}>
                <div className="relative">
                    <img src={`/img/planet-${getRandomImageId()}.svg`}
                    alt={`Image of planet ${planet.name}`}
                    title={`Image of planet ${planet.name}`}
                    className={`
                        block my-4 mx-auto w-[200px]
                    `}
                    />
                    <div className={`
                        absolute top-0 bottom-0 left-0 right-0
                        flex align-middle items-center justify-around	
                        opacity-0
                        hover:opacity-100 
                        `
                    }>
                        <Button>
                                {`show more`.toUpperCase()}
                        </Button>
                    </div>
                </div>
            </Link>
        )
    }
}
