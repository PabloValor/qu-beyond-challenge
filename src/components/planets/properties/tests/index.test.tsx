import { MemoryRouter } from "react-router-dom";
import { ReactTestRenderer, ReactTestRendererJSON, create } from "react-test-renderer";
import { render } from "@testing-library/react";
import PlanetProperties from "..";
import { Planet as TPlanet } from '../../../../types/planet'


const planetMock: TPlanet = {
    id: "123",
    name: "Earth",
    diameter: "12742",
    rotation_period: "24",
    orbital_period: "365",
    gravity: "9.8",
    population: "7800000000",
    climate: "Temperate",
    terrain: "Land, Water",
    surface_water: "70",
    residents: [],
    films: [],
    url: "https://example.com/earth",
    created: "2021-01-01",
    edited: "2021-01-02",
}

function toJson(component: ReactTestRenderer) {
    const result = component.toJSON()
    expect(result).toBeDefined()
    return result as ReactTestRendererJSON
    }
    
    describe('<PlanetProperties />', () => {
    
        test('should match snapshot', () => { 
            const component = create(
            <MemoryRouter>
                <PlanetProperties planet={planetMock}/>
            </MemoryRouter>
            )
            let json = toJson(component)
            expect(json).toMatchSnapshot() 
        })
    
        test('PlanetProperties mounts properly', () => {
            const wrapper = render(<MemoryRouter> <PlanetProperties planet={planetMock}/></MemoryRouter> )
            expect(wrapper).toBeTruthy()
    
            const span = wrapper.container.querySelector('span')
            expect(span?.textContent).toBe(` Diameter: ${planetMock.diameter}`)
        })
    });