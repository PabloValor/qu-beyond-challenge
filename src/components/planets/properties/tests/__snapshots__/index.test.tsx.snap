// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<PlanetProperties /> > should match snapshot 1`] = `
<div
  className="block md:flex place-content-evenly gap-4 mb-4"
>
  <div
    className="bg-slate-300 p-4 rounded-lg shadow-lg w-full mb-3 md:mb-0"
  >
    <div
      className="uppercase my-4"
    >
      <span
        className="bg-slate-50 p-2 mr-1 rounded-md font-bold"
      >
        <svg
          className="inline-block text-2xl mr-1"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M418.3 80H307.8c-7.5 0-13.7 6.1-13.7 13.7v.2c0 7.5 6.1 13.7 13.7 13.7h76.5L107.6 384.3v-76.5c0-7.5-6.1-13.7-13.7-13.7h-.2c-7.5 0-13.7 6.1-13.7 13.7v110.5c0 7.5 6.1 13.7 13.7 13.7h110.5c7.5 0 13.7-6.1 13.7-13.7v-.2c0-7.5-6.1-13.7-13.7-13.7h-76.5l276.7-276.7v76.5c0 7.5 6.1 13.7 13.7 13.7h.2c7.5 0 13.7-6.1 13.7-13.7V93.7c0-7.6-6.1-13.7-13.7-13.7z"
          />
        </svg>
         Diameter: 
        <span
          className="font-bold"
        >
          12742
        </span>
      </span>
    </div>
    <div
      className="uppercase my-4"
    >
      <span
        className="bg-slate-50 p-2 mr-1 rounded-md font-bold"
      >
        <svg
          className="inline-block mr-1 text-2xl"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M417.063 85.625c-35.503-.147-80.717 9.822-129.563 28.97-9.31-1.8-18.804-2.706-28.344-2.69-25.04.045-50.414 6.454-73.656 19.907-32.044 18.55-54.554 47.287-65.813 79.782-80.284 64.16-123.395 133.9-100.718 173.28 23.35 40.55 109.384 36.656 208.593-4.218 1.7-.7 3.417-1.403 5.125-2.125.622-.262 1.25-.514 1.875-.78 1.81-.765 3.62-1.554 5.437-2.344l1.563-.687c1.443-.633 2.895-1.29 4.343-1.94 19.712-8.812 39.79-19.016 59.844-30.624 15.83-9.162 30.907-18.77 45.156-28.656.394-.273.795-.54 1.188-.813.672-.465 1.33-.94 2-1.406 100.373-70.007 158.15-152.83 132.625-197.155-11.013-19.123-35.953-28.36-69.658-28.5zm-56.375 50c24.466-.44 42.61 5.846 50.437 19.438 8.373 14.54 3.594 35.145-11.22 57.937-3.086-9.146-7.093-18.13-12.092-26.813-11.72-20.35-27.54-36.846-45.782-49.093 6.543-.87 12.788-1.364 18.658-1.47zM111.75 269.905c1.397 22.12 7.76 44.257 19.563 64.75 5.088 8.838 10.948 16.957 17.437 24.314-27.663 1.7-48.293-4.418-56.78-19.158-9.82-17.05-1.558-42.433 19.78-69.906zm291.22 27.658c-25.755 21.322-55.33 41.912-87.876 60.75-31.97 18.504-63.966 33.663-94.75 45.28 36.843 10.142 77.578 6.073 113.28-14.593 35.833-20.74 59.75-54.215 69.345-91.438z"
          />
        </svg>
         Rotation Period: 
        <span
          className="font-bold"
        >
          24
        </span>
      </span>
    </div>
    <div
      className="uppercase my-4"
    >
      <span
        className="bg-slate-50 p-2 mr-1 rounded-md font-bold"
      >
        <svg
          className="inline-block mr-1 text-2xl"
          fill="none"
          height="1em"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="3"
          />
          <circle
            cx="18"
            cy="6"
            r="2"
          />
          <circle
            cx="6"
            cy="18"
            r="2"
          />
          <path
            d="M12 20a8 8 0 0 0 6.82-12.17"
          />
          <path
            d="M12 4a8 8 0 0 0-6.82 12.17"
          />
        </svg>
         Orbital Period: 
        <span
          className="font-bold"
        >
          365
        </span>
      </span>
    </div>
    <div
      className="uppercase my-4"
    >
      <span
        className="bg-slate-50 p-2 mr-1 rounded-md font-bold"
      >
        <svg
          className="inline-block mr-1 text-2xl"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M301.1 212c4.4 4.4 4.4 11.9 0 16.3l-9.7 9.7c-4.4 4.7-11.9 4.7-16.6 0l-10.5-10.5c-4.4-4.7-4.4-11.9 0-16.6l9.7-9.7c4.4-4.4 11.9-4.4 16.6 0l10.5 10.8zm-30.2-19.7c3-3 3-7.8 0-10.5-2.8-3-7.5-3-10.5 0-2.8 2.8-2.8 7.5 0 10.5 3.1 2.8 7.8 2.8 10.5 0zm-26 5.3c-3 2.8-3 7.5 0 10.2 2.8 3 7.5 3 10.5 0 2.8-2.8 2.8-7.5 0-10.2-3-3-7.7-3-10.5 0zm72.5-13.3c-19.9-14.4-33.8-43.2-11.9-68.1 21.6-24.9 40.7-17.2 59.8.8 11.9 11.3 29.3 24.9 17.2 48.2-12.5 23.5-45.1 33.2-65.1 19.1zm47.7-44.5c-8.9-10-23.3 6.9-15.5 16.1 7.4 9 32.1 2.4 15.5-16.1zM504 256c0 137-111 248-248 248S8 393 8 256 119 8 256 8s248 111 248 248zm-66.2 42.6c2.5-16.1-20.2-16.6-25.2-25.7-13.6-24.1-27.7-36.8-54.5-30.4 11.6-8 23.5-6.1 23.5-6.1.3-6.4 0-13-9.4-24.9 3.9-12.5.3-22.4.3-22.4 15.5-8.6 26.8-24.4 29.1-43.2 3.6-31-18.8-59.2-49.8-62.8-22.1-2.5-43.7 7.7-54.3 25.7-23.2 40.1 1.4 70.9 22.4 81.4-14.4-1.4-34.3-11.9-40.1-34.3-6.6-25.7 2.8-49.8 8.9-61.4 0 0-4.4-5.8-8-8.9 0 0-13.8 0-24.6 5.3 11.9-15.2 25.2-14.4 25.2-14.4 0-6.4-.6-14.9-3.6-21.6-5.4-11-23.8-12.9-31.7 2.8.1-.2.3-.4.4-.5-5 11.9-1.1 55.9 16.9 87.2-2.5 1.4-9.1 6.1-13 10-21.6 9.7-56.2 60.3-56.2 60.3-28.2 10.8-77.2 50.9-70.6 79.7.3 3 1.4 5.5 3 7.5-2.8 2.2-5.5 5-8.3 8.3-11.9 13.8-5.3 35.2 17.7 24.4 15.8-7.2 29.6-20.2 36.3-30.4 0 0-5.5-5-16.3-4.4 27.7-6.6 34.3-9.4 46.2-9.1 8 3.9 8-34.3 8-34.3 0-14.7-2.2-31-11.1-41.5 12.5 12.2 29.1 32.7 28 60.6-.8 18.3-15.2 23-15.2 23-9.1 16.6-43.2 65.9-30.4 106 0 0-9.7-14.9-10.2-22.1-17.4 19.4-46.5 52.3-24.6 64.5 26.6 14.7 108.8-88.6 126.2-142.3 34.6-20.8 55.4-47.3 63.9-65 22 43.5 95.3 94.5 101.1 59z"
          />
        </svg>
         Gravity:  
        <span
          className="font-bold"
        >
          9.8
        </span>
      </span>
    </div>
    <div
      className="uppercase my-4"
    >
      <span
        className="bg-slate-50 p-2 mr-1 rounded-md font-bold"
      >
        <svg
          className="inline-block mr-1 text-2xl"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M337.454 232c33.599 0 61.092-27.002 61.092-60 0-32.997-27.493-60-61.092-60s-61.09 27.003-61.09 60c0 32.998 27.491 60 61.09 60zm-162.908 0c33.599 0 61.09-27.002 61.09-60 0-32.997-27.491-60-61.09-60s-61.092 27.003-61.092 60c0 32.998 27.493 60 61.092 60zm0 44C126.688 276 32 298.998 32 346v54h288v-54c0-47.002-97.599-70-145.454-70zm162.908 11.003c-6.105 0-10.325 0-17.454.997 23.426 17.002 32 28 32 58v54h128v-54c0-47.002-94.688-58.997-142.546-58.997z"
          />
        </svg>
         Population: 
        <span
          className="font-bold"
        >
          7800000000
        </span>
      </span>
    </div>
  </div>
  <div
    className="bg-slate-300 p-4 rounded-lg shadow-lg w-full"
  >
    <div
      className="uppercase my-4"
    >
      <span
        className="bg-slate-50 p-2 mr-1 rounded-md font-bold"
      >
        <svg
          baseProfile="tiny"
          className="inline-block mr-1 text-2xl"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          version="1.2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              d="M15 22c-.552 0-1-.447-1-1v-6c0-.553.448-1 1-1s1 .447 1 1v6c0 .553-.448 1-1 1zM9 22c-.552 0-1-.447-1-1v-6c0-.553.448-1 1-1s1 .447 1 1v6c0 .553-.448 1-1 1zM12 24c-.552 0-1-.447-1-1v-6c0-.553.448-1 1-1s1 .447 1 1v6c0 .553-.448 1-1 1zM6 18c-2.206 0-4-1.794-4-4 0-1.861 1.277-3.429 3.001-3.874l-.001-.126c0-3.309 2.691-6 6-6 2.587 0 4.824 1.639 5.65 4.015 2.936-.244 5.35 2.113 5.35 4.985 0 2.241-1.507 4.223-3.666 4.819-.535.146-1.083-.166-1.23-.697-.147-.532.165-1.083.698-1.23 1.294-.358 2.198-1.547 2.198-2.892 0-1.654-1.346-3-3-3-.242 0-.499.041-.811.13l-1.074.306-.185-1.102c-.326-1.932-1.978-3.334-3.93-3.334-2.206 0-4 1.794-4 4 0 .272.027.545.082.808l.248 1.202-1.422-.016c-1.011.006-1.908.903-1.908 2.006s.897 2 2 2c.552 0 1 .447 1 1s-.448 1-1 1z"
            />
          </g>
        </svg>
        Climate: 
        <span
          className="font-bold"
        >
          Temperate
        </span>
      </span>
    </div>
    <div
      className="uppercase my-4"
    >
      <span
        className="bg-slate-50 p-2 mr-1 rounded-md font-bold"
      >
        <svg
          className="inline-block mr-1 text-2xl"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M14 6l-3.75 5 2.85 3.8-1.6 1.2C9.81 13.75 7 10 7 10l-6 8h22L14 6z"
          />
        </svg>
        Terrain: 
        <span
          className="font-bold"
        >
          Land, Water
        </span>
      </span>
    </div>
    <div
      className="uppercase my-4"
    >
      <span
        className="bg-slate-50 p-2 mr-1 rounded-md font-bold"
      >
        <svg
          className="inline-block mr-1 text-2xl"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 576 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M562.1 383.9c-21.5-2.4-42.1-10.5-57.9-22.9-14.1-11.1-34.2-11.3-48.2 0-37.9 30.4-107.2 30.4-145.7-1.5-13.5-11.2-33-9.1-46.7 1.8-38 30.1-106.9 30-145.2-1.7-13.5-11.2-33.3-8.9-47.1 2-15.5 12.2-36 20.1-57.7 22.4-7.9.8-13.6 7.8-13.6 15.7v32.2c0 9.1 7.6 16.8 16.7 16 28.8-2.5 56.1-11.4 79.4-25.9 56.5 34.6 137 34.1 192 0 56.5 34.6 137 34.1 192 0 23.3 14.2 50.9 23.3 79.1 25.8 9.1.8 16.7-6.9 16.7-16v-31.6c.1-8-5.7-15.4-13.8-16.3zm0-144c-21.5-2.4-42.1-10.5-57.9-22.9-14.1-11.1-34.2-11.3-48.2 0-37.9 30.4-107.2 30.4-145.7-1.5-13.5-11.2-33-9.1-46.7 1.8-38 30.1-106.9 30-145.2-1.7-13.5-11.2-33.3-8.9-47.1 2-15.5 12.2-36 20.1-57.7 22.4-7.9.8-13.6 7.8-13.6 15.7v32.2c0 9.1 7.6 16.8 16.7 16 28.8-2.5 56.1-11.4 79.4-25.9 56.5 34.6 137 34.1 192 0 56.5 34.6 137 34.1 192 0 23.3 14.2 50.9 23.3 79.1 25.8 9.1.8 16.7-6.9 16.7-16v-31.6c.1-8-5.7-15.4-13.8-16.3zm0-144C540.6 93.4 520 85.4 504.2 73 490.1 61.9 470 61.7 456 73c-37.9 30.4-107.2 30.4-145.7-1.5-13.5-11.2-33-9.1-46.7 1.8-38 30.1-106.9 30-145.2-1.7-13.5-11.2-33.3-8.9-47.1 2-15.5 12.2-36 20.1-57.7 22.4-7.9.8-13.6 7.8-13.6 15.7v32.2c0 9.1 7.6 16.8 16.7 16 28.8-2.5 56.1-11.4 79.4-25.9 56.5 34.6 137 34.1 192 0 56.5 34.6 137 34.1 192 0 23.3 14.2 50.9 23.3 79.1 25.8 9.1.8 16.7-6.9 16.7-16v-31.6c.1-8-5.7-15.4-13.8-16.3z"
          />
        </svg>
        Surface water: 
        <span
          className="font-bold"
        >
          70
        </span>
      </span>
    </div>
  </div>
</div>
`;
